internal/backend/db/dbgorm/triggers.go

```diff
@@ -14,24 +14,33 @@ import (
 
 func triggerToRecord(ctx context.Context, tx *tx, trigger sdktypes.Trigger) (*scheme.Trigger, error) {
 	connID := trigger.ConnectionID()
+	var conn sdktypes.Connection
+	var err error
+	var projID sdktypes.ProjectID
 
-	conn, err := tx.GetConnection(ctx, connID)
-	if err != nil {
-		return nil, fmt.Errorf("get trigger connection: %w", err)
+	if connID.IsValid() {
+		conn, err = tx.GetConnection(ctx, connID)
+		if err != nil {
+			return nil, fmt.Errorf("get trigger connection: %w", err)
+		}
+		projID = conn.ProjectID()
 	}
 
-	projID := conn.ProjectID()
-
 	envID := trigger.EnvID()
 	if envID.IsValid() {
 		env, err := tx.GetEnvByID(ctx, envID)
 		if err != nil {
 			return nil, fmt.Errorf("get trigger env: %w", err)
 		}
 
-		if projID != env.ProjectID() {
+		if projID.IsValid() && projID != env.ProjectID() {
 			return nil, fmt.Errorf("env and connection project mismatch: %v != %v", projID, env.ProjectID())
 		}
+		projID = env.ProjectID()
+	}
+
+	if !projID.IsValid() {
```

Comment:
move this check after line 26 and fail fast

then you can remove the projID.IsValid() in line 36 as well